## 2. Imports & File Paths

In [1]:
# import necessary libraries

import pandas as pd
import json
import os




In [2]:
# Set base data path
BASE_PATH = "../data/raw/"  # adjust if notebook sits differently

## 3. Load the raw data 

## 3a. Load Transactions Data

In [3]:
transactions_path = os.path.join(BASE_PATH, "transactions_data.csv")

transactions = pd.read_csv(transactions_path)
print("Transactions shape:", transactions.shape)
transactions.head()


Transactions shape: (13305915, 12)


,id,date,client_id,card_id,amount,use_chip,merchant_id,merchant_city,merchant_state,zip,mcc,errors
0,7475327,2010-01-01 00:01:00,1556,2972,$-77.00,Swipe Transaction,59935,Beulah,ND,58523.0,5499,NaN
1,7475328,2010-01-01 00:02:00,561,4575,$14.57,Swipe Transaction,67570,Bettendorf,IA,52722.0,5311,NaN
2,7475329,2010-01-01 00:02:00,1129,102,$80.00,Swipe Transaction,27092,Vista,CA,92084.0,4829,NaN
3,7475331,2010-01-01 00:05:00,430,2860,$200.00,Swipe Transaction,27092,Crown Point,IN,46307.0,4829,NaN
4,7475332,2010-01-01 00:06:00,848,3915,$46.41,Swipe Transaction,13051,Harwood,MD,20776.0,5813,NaN


In [4]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13305915 entries, 0 to 13305914
Data columns (total 12 columns):
 #   Column          Dtype  
---  ------          -----  
 0   id              int64  
 1   date            object 
 2   client_id       int64  
 3   card_id         int64  
 4   amount          object 
 5   use_chip        object 
 6   merchant_id     int64  
 7   merchant_city   object 
 8   merchant_state  object 
 9   zip             float64
 10  mcc             int64  
 11  errors          object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.2+ GB


3b. Load Card Data

In [5]:
cards_path = os.path.join(BASE_PATH, "cards_data.csv")

cards = pd.read_csv(cards_path)
print("Cards shape:", cards.shape)
cards.head()


Cards shape: (6146, 13)


,id,client_id,card_brand,card_type,card_number,expires,cvv,has_chip,num_cards_issued,credit_limit,acct_open_date,year_pin_last_changed,card_on_dark_web
0,4524,825,Visa,Debit,4344676511950444,12/2022,623,YES,2,$24295,09/2002,2008,No
1,2731,825,Visa,Debit,4956965974959986,12/2020,393,YES,2,$21968,04/2014,2014,No
2,3701,825,Visa,Debit,4582313478255491,02/2024,719,YES,2,$46414,07/2003,2004,No
3,42,825,Visa,Credit,4879494103069057,08/2024,693,NO,1,$12400,01/2003,2012,No
4,4659,825,Mastercard,Debit (Prepaid),5722874738736011,03/2009,75,YES,1,$28,09/2008,2009,No


In [6]:
cards.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6146 entries, 0 to 6145
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   id                     6146 non-null   int64 
 1   client_id              6146 non-null   int64 
 2   card_brand             6146 non-null   object
 3   card_type              6146 non-null   object
 4   card_number            6146 non-null   int64 
 5   expires                6146 non-null   object
 6   cvv                    6146 non-null   int64 
 7   has_chip               6146 non-null   object
 8   num_cards_issued       6146 non-null   int64 
 9   credit_limit           6146 non-null   object
 10  acct_open_date         6146 non-null   object
 11  year_pin_last_changed  6146 non-null   int64 
 12  card_on_dark_web       6146 non-null   object
dtypes: int64(6), object(7)
memory usage: 624.3+ KB


3c. Load User Data

In [7]:
users_path = os.path.join(BASE_PATH, "users_data.csv")

users = pd.read_csv(users_path)
print("Users shape:", users.shape)
users.head()


Users shape: (2000, 14)


,id,current_age,retirement_age,birth_year,birth_month,gender,address,latitude,longitude,per_capita_income,yearly_income,total_debt,credit_score,num_credit_cards
0,825,53,66,1966,11,Female,462 Rose Lane,34.15,-117.76,$29278,$59696,$127613,787,5
1,1746,53,68,1966,12,Female,3606 Federal Boulevard,40.76,-73.74,$37891,$77254,$191349,701,5
2,1718,81,67,1938,11,Female,766 Third Drive,34.02,-117.89,$22681,$33483,$196,698,5
3,708,63,63,1957,1,Female,3 Madison Street,40.71,-73.99,$163145,$249925,$202328,722,4
4,1164,43,70,1976,9,Male,9620 Valley Stream Drive,37.76,-122.44,$53797,$109687,$183855,675,1


In [8]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 2000 non-null   int64  
 1   current_age        2000 non-null   int64  
 2   retirement_age     2000 non-null   int64  
 3   birth_year         2000 non-null   int64  
 4   birth_month        2000 non-null   int64  
 5   gender             2000 non-null   object 
 6   address            2000 non-null   object 
 7   latitude           2000 non-null   float64
 8   longitude          2000 non-null   float64
 9   per_capita_income  2000 non-null   object 
 10  yearly_income      2000 non-null   object 
 11  total_debt         2000 non-null   object 
 12  credit_score       2000 non-null   int64  
 13  num_credit_cards   2000 non-null   int64  
dtypes: float64(2), int64(7), object(5)
memory usage: 218.9+ KB


3d. Load Fraud Labels

In [9]:
labels_path = os.path.join(BASE_PATH, "train_fraud_labels.json")

with open(labels_path, "r") as f:
    labels_json = json.load(f)

labels = pd.DataFrame(labels_json)
print("Labels shape:", labels.shape)
labels.head()


Labels shape: (8914963, 1)


,target
10000000,No
10000002,No
10000003,No
10000004,No
10000005,No


In [10]:

labels.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8914963 entries, 10000000 to 9999998
Data columns (total 1 columns):
 #   Column  Dtype 
---  ------  ----- 
 0   target  object
dtypes: object(1)
memory usage: 136.0+ MB


In [11]:
print(type(labels_json))
print(len(labels_json))

<class 'dict'>
1


In [12]:
list(labels_json.keys())


['target']

In [13]:
inner = next(iter(labels_json.values()))
print(type(inner))


<class 'dict'>


In [14]:
len(inner)
first_inner_key = next(iter(inner))
first_inner_key, inner[first_inner_key]


('10649266', 'No')

3e. Load MCC Codes (Merchant Category Codes)

In [15]:
mcc_path = os.path.join(BASE_PATH, "mcc_codes.json")

with open(mcc_path, "r") as f:
    mcc_json = json.load(f)

type(mcc_json), len(mcc_json)





(dict, 109)

In [16]:
list(mcc_json.items())[:5]


[('5812', 'Eating Places and Restaurants'),
 ('5541', 'Service Stations'),
 ('7996', 'Amusement Parks, Carnivals, Circuses'),
 ('5411', 'Grocery Stores, Supermarkets'),
 ('4784', 'Tolls and Bridge Fees')]

In [17]:
mcc_items = list(mcc_json.items())

mcc = pd.DataFrame(mcc_items, columns=["merchant_category_code", "mcc_description"])
print("MCC shape:", mcc.shape)
mcc.head()


MCC shape: (109, 2)


,merchant_category_code,mcc_description
0,5812,Eating Places and Restaurants
1,5541,Service Stations
2,7996,"Amusement Parks, Carnivals, Circuses"
3,5411,"Grocery Stores, Supermarkets"
4,4784,Tolls and Bridge Fees


In [18]:
mcc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 109 entries, 0 to 108
Data columns (total 2 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   merchant_category_code  109 non-null    object
 1   mcc_description         109 non-null    object
dtypes: object(2)
memory usage: 1.8+ KB


3f. Inspect Columns of Each Table

In [19]:
print("Transactions columns:\n", transactions.columns.tolist(), "\n")
print("Cards columns:\n", cards.columns.tolist(), "\n")
print("Users columns:\n", users.columns.tolist(), "\n")
print("Labels columns:\n", labels.columns.tolist(), "\n")
print("MCC columns:\n", mcc.columns.tolist(), "\n")


Transactions columns:
 ['id', 'date', 'client_id', 'card_id', 'amount', 'use_chip', 'merchant_id', 'merchant_city', 'merchant_state', 'zip', 'mcc', 'errors'] 

Cards columns:
 ['id', 'client_id', 'card_brand', 'card_type', 'card_number', 'expires', 'cvv', 'has_chip', 'num_cards_issued', 'credit_limit', 'acct_open_date', 'year_pin_last_changed', 'card_on_dark_web'] 

Users columns:
 ['id', 'current_age', 'retirement_age', 'birth_year', 'birth_month', 'gender', 'address', 'latitude', 'longitude', 'per_capita_income', 'yearly_income', 'total_debt', 'credit_score', 'num_credit_cards'] 

Labels columns:
 ['target'] 

MCC columns:
 ['merchant_category_code', 'mcc_description'] 



## 4. Understanding the Data

4a. missing values check

In [20]:
transactions.isnull().sum()

id                       0
date                     0
client_id                0
card_id                  0
amount                   0
use_chip                 0
merchant_id              0
merchant_city            0
merchant_state     1563700
zip                1652706
mcc                      0
errors            13094522
dtype: int64

In [21]:
transactions.isna().mean().sort_values(ascending=False).head(10)


errors            0.984113
zip               0.124208
merchant_state    0.117519
id                0.000000
date              0.000000
client_id         0.000000
card_id           0.000000
amount            0.000000
use_chip          0.000000
merchant_id       0.000000
dtype: float64

In [22]:
cards.isna().mean().sort_values(ascending=False).head(10)

id                  0.0
client_id           0.0
card_brand          0.0
card_type           0.0
card_number         0.0
expires             0.0
cvv                 0.0
has_chip            0.0
num_cards_issued    0.0
credit_limit        0.0
dtype: float64

In [23]:
users.isna().mean().sort_values(ascending=False).head(10)

id                   0.0
current_age          0.0
retirement_age       0.0
birth_year           0.0
birth_month          0.0
gender               0.0
address              0.0
latitude             0.0
longitude            0.0
per_capita_income    0.0
dtype: float64

4b. Check duplicates on key IDs

In [24]:
transactions["id"].is_unique, transactions["id"].nunique(), len(transactions)

(True, 13305915, 13305915)

In [25]:
cards["id"].is_unique, cards["id"].nunique(), len(cards)

(True, 6146, 6146)

In [26]:
users["id"].is_unique, users["id"].nunique(), len(users)

(True, 2000, 2000)

In [27]:
cards["client_id"].nunique()

2000

In [28]:
users["id"].nunique()

2000

In [29]:
transactions["client_id"].nunique()

1219

In [30]:
(True, 13305915, 13305915)



(True, 13305915, 13305915)

Data merging

Align with transactions

From transaction.info() output:

In transactions, the transaction ID column is id (int64, unique).

the labels use transaction_id (int64, from JSON keys).

So before merging, rename in transactions:

Clean & rename columns

Transactions

In [31]:
transactions = transactions.rename(columns={"id": "transaction_id"})


In [32]:
transactions = transactions.rename(
    columns={
        "id": "transaction_id",
        "date": "timestamp",
        "client_id": "user_id",
        "mcc": "merchant_category_code"
    }
)


Cards

In [33]:
# rename
cards = cards.rename(
    columns={
        "id": "card_id",
        "client_id": "user_id"
    }
)


Users

In [34]:
# rename
users = users.rename(
    columns={
        "id": "user_id"
    }
)


Build labels_df

In [35]:
# Build labels DataFrame properly
# inner is the dict: {transaction_id_str: "Yes"/"No"}
inner = next(iter(labels_json.values()))

# Turn into DataFrame
labels_df = pd.DataFrame(
    list(inner.items()),
    columns=["transaction_id_str", "target_str"]
)

labels_df.head()


,transaction_id_str,target_str
0,10649266,No
1,23410063,No
2,9316588,No
3,12478022,No
4,9558530,No


Convert to correct types

We want:

transaction_id as integer

target as 0/1 (0 = No, 1 = Yes)

In [36]:
labels_df["transaction_id"] = labels_df["transaction_id_str"].astype("int64")
labels_df["target"] = (labels_df["target_str"] == "Yes").astype("int8")

labels_df = labels_df[["transaction_id", "target"]]

labels_df.head()


,transaction_id,target
0,10649266,0
1,23410063,0
2,9316588,0
3,12478022,0
4,9558530,0



Merging;


In [37]:
tx_labeled = transactions.merge(labels_df, on="transaction_id", how="left")
print(tx_labeled.shape)
tx_labeled["target"].value_counts(dropna=False)


(13305915, 13)


target
0.0    8901631
NaN    4390952
1.0      13332
Name: count, dtype: int64

In [38]:
tx_labeled[["transaction_id", "target"]].head()

,transaction_id,target
0,7475327,0.0
1,7475328,0.0
2,7475329,0.0
3,7475331,NaN
4,7475332,0.0


In [39]:
transaction_labels_df= tx_labeled.dropna(subset=["target"]).copy()
transaction_labels_df["target"] = transaction_labels_df["target"].astype("int8")

print(transaction_labels_df.shape)

transaction_labels_df["target"].value_counts(normalize= True)


(8914963, 13)


target
0    0.998505
1    0.001495
Name: proportion, dtype: float64

In [40]:
transaction_labels__cards_df = transaction_labels_df.merge(cards, on= ["card_id", "user_id"], how= "left")
print("After cards merge:", transaction_labels__cards_df.shape)

After cards merge: (8914963, 24)


Reason for ["card_id", "user_id"] as join keys: a card belongs to a single user, and this makes mismatches less likely.

In [41]:
transaction_labels__cards_users_df = transaction_labels__cards_df.merge(users, on= "user_id", how= "left")
print("After users merge:", transaction_labels__cards_users_df.shape)

After users merge: (8914963, 37)


Merge mcc into transaction_labels__cards_users_df

In [42]:
# convert both dtypes as object
transaction_labels__cards_users_df["merchant_category_code"] = transaction_labels__cards_users_df["merchant_category_code"].astype(str)
mcc["merchant_category_code"] = mcc["merchant_category_code"].astype(str)


In [43]:
# merge
all_merged_df = transaction_labels__cards_users_df.merge(mcc, on="merchant_category_code", how="left")
print("After mcc merge:", all_merged_df.shape)


After mcc merge: (8914963, 38)


In [44]:
all_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8914963 entries, 0 to 8914962
Data columns (total 38 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   transaction_id          int64  
 1   timestamp               object 
 2   user_id                 int64  
 3   card_id                 int64  
 4   amount                  object 
 5   use_chip                object 
 6   merchant_id             int64  
 7   merchant_city           object 
 8   merchant_state          object 
 9   zip                     float64
 10  merchant_category_code  object 
 11  errors                  object 
 12  target                  int8   
 13  card_brand              object 
 14  card_type               object 
 15  card_number             int64  
 16  expires                 object 
 17  cvv                     int64  
 18  has_chip                object 
 19  num_cards_issued        int64  
 20  credit_limit            object 
 21  acct_open_date          object 

Basic cleaning on merged table

In [45]:
# Convert timestamp to datetime
all_merged_df["timestamp"] = pd.to_datetime(all_merged_df["timestamp"])

In [46]:


#  Convert amount & numeric-like strings to float
for col in ["amount", "credit_limit", "per_capita_income", "yearly_income", "total_debt"]:
    all_merged_df[col] = (
        all_merged_df[col]
        .astype(str)
        .str.replace(",", "", regex=False)
        .str.replace("$", "", regex=False)  # if present
        .astype(float)
    )

In [47]:
all_merged_df.isnull().sum()

transaction_id                  0
timestamp                       0
user_id                         0
card_id                         0
amount                          0
use_chip                        0
merchant_id                     0
merchant_city                   0
merchant_state            1047865
zip                       1107377
merchant_category_code          0
errors                    8773196
target                          0
card_brand                      0
card_type                       0
card_number                     0
expires                         0
cvv                             0
has_chip                        0
num_cards_issued                0
credit_limit                    0
acct_open_date                  0
year_pin_last_changed           0
card_on_dark_web                0
current_age                     0
retirement_age                  0
birth_year                      0
birth_month                     0
gender                          0
address       

In [48]:
# handling missing values in merchant_state

# Fill missing merchant_state with "ONLINE" only when merchant_city is ONLINE.
# The way the dataset works: if merchant_city is 'ONLINE' then the merchant_state is blank, hence will append with merchant_state as ONLINE. 
# Fraud detection models may benefit from distinguishing between online & offline transactions.
all_merged_df.loc[all_merged_df['merchant_city'] == 'ONLINE', 'merchant_state'] = 'ONLINE'


In [49]:
import numpy as np

In [50]:
# Imputing missing ZIP codes using merchant-based mode imputation.
# from the data dictionary, Since ZIP codes correspond to merchants (not users), I will fill missing values
# using the most frequent ZIP per `merchant_id`. If a merchant has no known ZIPs,
# use the most common ZIP in the dataset as a fallback.

# Fill missing ZIPs with the most frequent ZIP per merchant_id
all_merged_df['zip'] = all_merged_df.groupby('merchant_id')['zip'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else np.nan))
# Fill remaining missing ZIPs with the most frequent ZIP in the dataset
all_merged_df['zip'].fillna(all_merged_df['zip'].mode()[0], inplace=True)
# Convert ZIP to integer format
all_merged_df['zip'] = all_merged_df['zip'].astype(int)

In [51]:
# 3. Drop obvious columns that are not useful for modeling
drop_cols = [
    "errors",          # mostly NaN
    "card_number",     # VERY high-cardinality + leaks sensitive info
    "cvv",             # sensitive & not useful
    "address"          # high-cardinality free text
]

cols_to_drop = [c for c in drop_cols if c in all_merged_df.columns]

all_merged_df = all_merged_df.drop(columns=cols_to_drop)

all_merged_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8914963 entries, 0 to 8914962
Data columns (total 34 columns):
 #   Column                  Dtype         
---  ------                  -----         
 0   transaction_id          int64         
 1   timestamp               datetime64[ns]
 2   user_id                 int64         
 3   card_id                 int64         
 4   amount                  float64       
 5   use_chip                object        
 6   merchant_id             int64         
 7   merchant_city           object        
 8   merchant_state          object        
 9   zip                     int32         
 10  merchant_category_code  object        
 11  target                  int8          
 12  card_brand              object        
 13  card_type               object        
 14  expires                 object        
 15  has_chip                object        
 16  num_cards_issued        int64         
 17  credit_limit            float64       
 18  ac

In [52]:
# check unique values in gender
all_merged_df["gender"].unique()

array(['Female', 'Male'], dtype=object)

In [53]:
all_merged_df["gender"]= all_merged_df["gender"].str.strip().str.lower()

In [54]:
print(all_merged_df["gender"].unique())

['female' 'male']


In [55]:
all_merged_df["use_chip"].unique()

array(['Swipe Transaction', 'Online Transaction', 'Chip Transaction'],
      dtype=object)

In [57]:
# Create transaction_hour first from timestamp
all_merged_df['transaction_hour'] = all_merged_df['timestamp'].dt.hour


In [58]:
# create day of week for transaction from timestamp
all_merged_df['transaction_dayofweek'] = all_merged_df['timestamp'].dt.dayofweek


In [69]:
# ---- Amount checks ----
non_positive_mask = all_merged_df['amount'] <= 0
amount_p99 = all_merged_df['amount'].quantile(0.99)
amount_p999 = all_merged_df['amount'].quantile(0.999)

print("===== AMOUNT CHECKS =====")
print("Non-positive amounts:", non_positive_mask.sum())
print("Amount P99:", amount_p99)
print("Amount P999:", amount_p999)

===== AMOUNT CHECKS =====
Non-positive amounts: 449972
Amount P99: 315.86
Amount P999: 890.3922800000385


In [72]:
# Non-positive amount flag
all_merged_df['amount_non_positive_flag'] = (all_merged_df['amount'] <= 0).astype(int)

About 449,972 transactions (~5%) have non-positive amounts, which likely correspond to refunds, reversals, or test behavior and should be kept and analyzed, not dropped. The 99th percentile is ~315.86 and the 99.9th percentile is ~890.39, so amounts above ~900 are rare and high-value, and should be treated as high-risk but informative rather than removed.

In [70]:
age_out_of_range_mask = (all_merged_df['current_age'] < 18) | (all_merged_df['current_age'] > 100)
computed_age = all_merged_df['timestamp'].dt.year - all_merged_df['birth_year']
age_mismatch_mask = (abs(all_merged_df['current_age'] - computed_age) > 1)

print("\n===== AGE CHECKS =====")
print("Age out of range:", age_out_of_range_mask.sum())
print("Age mismatch (>1 year):", age_mismatch_mask.sum())



===== AGE CHECKS =====
Age out of range: 10305
Age mismatch (>1 year): 7373712


About 10,305 records have ages outside the expected 18–100 range, which is small and likely due to data entry or profile inconsistencies. However, 7.37M records (~83%) show a >1-year mismatch between current_age and birth_year. This indicates that current_age is likely not updated regularly (static field), while birth_year is reliable. For modeling, computed_age should be used as the primary age feature.

In [73]:
#  Age out-of-range flag
all_merged_df['age_out_of_range_flag'] = (
    (all_merged_df['current_age'] < 18) | 
    (all_merged_df['current_age'] > 100)
).astype(int)

#  Age mismatch flag
all_merged_df['computed_age'] = all_merged_df['timestamp'].dt.year - all_merged_df['birth_year']
all_merged_df['age_mismatch_flag'] = (
    abs(all_merged_df['current_age'] - all_merged_df['computed_age']) > 1
).astype(int)

In [71]:
# ---- Geo / location checks ----
invalid_lat_mask = ~all_merged_df['latitude'].between(-90, 90)
invalid_lon_mask = ~all_merged_df['longitude'].between(-180, 180)

invalid_zips = ['00000', '99999', '12345']
zip_str = all_merged_df['zip'].astype(str).str.zfill(5)
invalid_zip_mask = (zip_str.isin(invalid_zips)) | (zip_str.str.len() != 5)

location_suspicious_mask = invalid_zip_mask | invalid_lat_mask | invalid_lon_mask

print("\n===== LOCATION CHECKS =====")
print("Invalid latitude:", invalid_lat_mask.sum())
print("Invalid longitude:", invalid_lon_mask.sum())
print("Suspicious ZIP codes:", invalid_zip_mask.sum())
print("Location suspicious:", location_suspicious_mask.sum())


===== LOCATION CHECKS =====
Invalid latitude: 0
Invalid longitude: 0
Suspicious ZIP codes: 0
Location suspicious: 0


All latitude, longitude, and ZIP values fall within valid ranges, and no placeholder or suspicious ZIP codes were detected. Location fields appear clean and consistent, with 0 invalid entries across all checks.


#  NORMALIZE STRING COLUMNS


str_cols = [
    'use_chip', 'has_chip', 'card_on_dark_web',
    'card_brand', 'card_type',
    'merchant_city', 'merchant_state',
    'gender', 'merchant_category_code', 'mcc_description'
]

for col in str_cols:
    if col in all_merged_df.columns:
        all_merged_df[col] = (
            all_merged_df[col]
            .astype(str)
            .str.strip()          # remove leading/trailing spaces
            .str.lower()          # normalize case
            .replace({'nan': np.nan})  # fix "nan" strings
        )

In [77]:
binary_cols = ['use_chip', 'has_chip', 'card_on_dark_web']

print("===== RAW BINARY VALUE CHECKS =====")
for col in binary_cols:
    if col in all_merged_df.columns:
        print(f"\n{col} unique values:")
        print(all_merged_df[col].value_counts(dropna=False))

===== RAW BINARY VALUE CHECKS =====

use_chip unique values:
use_chip
swipe transaction     4668212
chip transaction      3202776
online transaction    1043975
Name: count, dtype: int64

has_chip unique values:
has_chip
YES    8022614
NO      892349
Name: count, dtype: int64

card_on_dark_web unique values:
card_on_dark_web
No    8914963
Name: count, dtype: int64


In [79]:
# --- has_chip (binary YES/NO) ---
all_merged_df['has_chip'] = all_merged_df['has_chip'].str.lower().map({
    'yes': 1,
    'no': 0
})

# --- card_on_dark_web (binary but only "No") ---
all_merged_df['card_on_dark_web'] = all_merged_df['card_on_dark_web'].str.lower().map({
    'yes': 1,
    'no': 0
})



MemoryError: Unable to allocate 68.0 MiB for an array with shape (8914963,) and data type int64

### Data Quality checks before saving cleaned data

In [59]:
# check for duplicates
duplicates= all_merged_df[all_merged_df.duplicated()]
print(f"Total duplicate rows: {duplicates.shape[0]}")

Total duplicate rows: 0


In [60]:
# check missing values
all_merged_df.isnull().sum()

transaction_id            0
timestamp                 0
user_id                   0
card_id                   0
amount                    0
use_chip                  0
merchant_id               0
merchant_city             0
merchant_state            0
zip                       0
merchant_category_code    0
target                    0
card_brand                0
card_type                 0
expires                   0
has_chip                  0
num_cards_issued          0
credit_limit              0
acct_open_date            0
year_pin_last_changed     0
card_on_dark_web          0
current_age               0
retirement_age            0
birth_year                0
birth_month               0
gender                    0
latitude                  0
longitude                 0
per_capita_income         0
yearly_income             0
total_debt                0
credit_score              0
num_credit_cards          0
mcc_description           0
transaction_hour          0
transaction_dayofwee

In [61]:
all_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8914963 entries, 0 to 8914962
Data columns (total 36 columns):
 #   Column                  Dtype         
---  ------                  -----         
 0   transaction_id          int64         
 1   timestamp               datetime64[ns]
 2   user_id                 int64         
 3   card_id                 int64         
 4   amount                  float64       
 5   use_chip                object        
 6   merchant_id             int64         
 7   merchant_city           object        
 8   merchant_state          object        
 9   zip                     int32         
 10  merchant_category_code  object        
 11  target                  int8          
 12  card_brand              object        
 13  card_type               object        
 14  expires                 object        
 15  has_chip                object        
 16  num_cards_issued        int64         
 17  credit_limit            float64       
 18  ac

In [1]:
import pandas as pd

In [3]:
import pandas as pd
import numpy as np

# Set display options to show all columns and prevent truncation
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)
pd.set_option('display.max_rows', 50)
# Optional: Set max column width higher if values are long
pd.set_option('display.max_colwidth', 50) 

# Load the DataFrame
#df = pd.read_parquet(r"C:\Users\HP\Desktop\aisoc\fraud-detection-system\data\processed\all_merged.parquet")





In [ ]:
df= pd.read_parquet(r"C:\Users\HP\Desktop\aisoc\fraud-detection-system\data\processed\all_merged.parquet")
df.head()

In [4]:
print("--- 5 Random Samples Showing ALL 40 Features ---")
# Use .sample(5) to get 5 random rows for good variety
sample_df = df.sample(5) 
print(sample_df)

--- 5 Random Samples Showing ALL 40 Features ---
         transaction_id           timestamp  user_id  card_id  amount            use_chip  merchant_id merchant_city merchant_state    zip merchant_category_code  target  card_brand card_type  expires has_chip  num_cards_issued  credit_limit acct_open_date  year_pin_last_changed card_on_dark_web  current_age  retirement_age  birth_year  birth_month  gender  latitude  longitude  per_capita_income  yearly_income  total_debt  credit_score  num_credit_cards               mcc_description  transaction_hour  transaction_dayofweek  amount_non_positive_flag  age_out_of_range_flag  computed_age  age_mismatch_flag
246965          7916216 2010-04-22 06:13:00      973     3833    9.58   swipe transaction        19752          Como             MS  38619                   5411       0        Amex    Credit  09/2024      YES                 1        8300.0        07/2008                   2015               No           55              68        1964   

In [83]:
all_merged_df.to_parquet("../data/processed/all_merged.parquet")

In [84]:
all_merged_df.to_csv("../data/processed/all_merged_df.csv")